lib/joken_jwks/default_strategy_template.ex

```diff
@@ -200,33 +215,116 @@ defmodule JokenJwks.DefaultStrategyTemplate do
           |> Keyword.put(:jwks_url, url)
           |> Keyword.put(:telemetry_prefix, telemetry_prefix)
           |> Keyword.put(:jws_supported_algs, algs)
+          |> Keyword.put(:should_start, start?)
+          |> Keyword.put(:first_fetch_sync, first_fetch_sync)
 
-        do_init(start?, first_fetch_sync, opts)
+        # :strategy_name => atom
+        GenServer.start_link(__MODULE__, opts, name: opts[:strategy_name] || __MODULE__)
+      end
+
+      defp get_strategy_name(opts) do
+        (opts[:strategy_name] || __MODULE__)
+        |> case do
+          name when is_atom(name) -> name
+          {:via, Registry, {_registry_name, name}} when is_atom(name) -> Atom.to_string(name)
+          _ -> nil
+        end
+      end
+
+      defp get_ets_name(opts),
+        do: opts[:ets_name] || ets_name_with_strategy_name_prefixed(opts)
+
+      defp ets_name_with_strategy_name_prefixed(opts) do
+        opts
+        |> get_strategy_name()
+        |> case do
+          nil -> EtsCache
+          __MODULE__ -> EtsCache
+          name when is_binary(name) -> (name <> "_ets") |> String.to_atom()
+          name when is_atom(name) -> (Atom.to_string(name) <> "_ets") |> String.to_atom()
+        end
+      end
+
+      defp ets_func(func, args \\ []) do
+        strategy_name = strategy_name_by_pid(self())
+
+        opts = [strategy_name: strategy_name]
+        ets_name = get_ets_name(opts)
+        apply(EtsCache, func, [ets_name | args])
+      end
+
+      defp strategy_name_by_pid(pid) do
+        pid
+        |> JokenJwks.DynamicDefaultStrategyRegistry.lookup_name_by_pid()
+        |> case do
+          nil -> __MODULE__
+          name -> name
+        end
+      end
+
+      # Client
+      @impl SignerMatchStrategy
+      def match_signer_for_kid(pid, kid, opts) do
+        GenServer.call(pid, {:match_signer_for_kid, kid, opts})
+      end
+
+      # Server (callbacks)
+      @impl true
+      def init(opts) do
+        Process.flag(:trap_exit, true)
+        {should_start, opts} = Keyword.pop!(opts, :should_start)
+        {first_fetch_sync, opts} = Keyword.pop!(opts, :first_fetch_sync)
+        do_init(should_start, first_fetch_sync, opts)
       end
 
       defp do_init(should_start, first_fetch_sync, opts) do
         cond do
           should_start and first_fetch_sync ->
             fetch_signers(opts[:jwks_url], opts)
-            Task.start_link(__MODULE__, :poll, [opts])
+            {:ok, opts, {:continue, :start_poll}}
 
           should_start ->
-            {:ok, _} = start_fetch_signers(opts[:jwks_url], opts)
-            Task.start_link(__MODULE__, :poll, [opts])
+            fetch_signers(opts[:jwks_url], opts)
```

Comment:
:wave: we're currently trying to debug in our codebase related to `first_fetch_sync` and saw this change on `master`: Shouldn't the `!first_fetch_sync` case **not** include this line? Seems odd that the first two branches of this `cond` are the same.

Reply:
https://github.com/joken-elixir/joken_jwks/pull/39/files#diff-c971e4f3329485ad6a91eb81a345c2144a3b3eb6a035bad9f84395c1fcde24a0L267 in the original code also runs `fetch_signers` - but done in a `Task.async`.

It's been a while since I created the PR and I might have forgotten the nuances, but it seems I saw what was happening and decided that because of dynamically generated strategies (hence not really a need for doing Task.async), and to maintain parity with original code, `start_fetch_signers` was removed and the code inside it was used, hence `fetch_signers`. 

@victorolinasc correct me if I'm wrong - I might have missed a layer of understanding here.